plutus-core/testlib/PlutusCore/Generators/PIR.hs

```diff
@@ -0,0 +1,1752 @@
+genKindAndTypeWithCtx ctx = do
+  k <- arbitrary
+  runGenTm $ local (\ e -> e { geTypes = ctx }) ((k,) <$> genType k)
+
+-- CODE REVIEW: does this exist anywhere??
+-- | `substClosedType x sub ty` substitutes the closed type `sub` for `x` in `ty`
+substClosedType :: TyName -> Type TyName DefaultUni () -> Type TyName DefaultUni () -> Type TyName DefaultUni ()
+substClosedType x sub ty =
+  case ty of
+    TyVar _ y
+      | x == y    -> sub
+      | otherwise -> ty
+    TyFun _ a b   -> TyFun () (substClosedType x sub a) (substClosedType x sub b)
+    TyApp _ a b   -> TyApp () (substClosedType x sub a) (substClosedType x sub b)
+    TyLam _ y k b
+      | x == y    -> ty
+      | otherwise -> TyLam () y k $ substClosedType x sub b
+    TyForall _ y k b
+      | x == y    -> ty
+      | otherwise -> TyForall () y k $ substClosedType x sub b
+    TyBuiltin{}   -> ty
+    TyIFix{}      -> ty
+
+-- CODE REVIEW: does this exist anywhere?
+builtinKind :: SomeTypeIn DefaultUni -> Kind ()
+builtinKind (SomeTypeIn t) = case t of
+  DefaultUniProtoList -> Star :-> Star
+  DefaultUniProtoPair -> Star :-> Star :-> Star
+  DefaultUniApply f _ -> let _ :-> k = builtinKind (SomeTypeIn f) in k
+  _                   -> Star
+
+-- * Shrinking types and kinds
+
+-- | Shriking-order on kinds
+leKind :: Kind () -> Kind () -> Bool
+leKind k1 k2 = go (reverse $ args k1) (reverse $ args k2)
+  where
+    args Type{}            = []
+    args (KindArrow _ a b) = a : args b
+
+    go [] _                = True
+    go _ []                = False
+    go (k : ks) (k' : ks')
+      | leKind k k' = go ks ks'
+      | otherwise   = go (k : ks) ks'
+
+-- | Strict shrinking order on kinds
+ltKind :: Kind () -> Kind () -> Bool
+ltKind k k' = k /= k' && leKind k k'
+
+-- | Take a type in a context and a new target kind
+--   Precondition: new kind is smaller or equal to old kind of the type.
+--   TODO (later): also allow changing which context it's valid in
+fixKind :: HasCallStack
+        => Map TyName (Kind ())
+        -> Type TyName DefaultUni ()
+        -> Kind ()
+        -> Type TyName DefaultUni ()
+fixKind ctx ty k
+  -- Nothing to do if we already have the right kind
+  | inferKind_ ctx ty == k = ty
+  | not $ k `leKind` inferKind_ ctx ty =
+      error "fixKind not smaller"
+  | otherwise = case ty of
+    -- Switch a variable out for a different variable of the right kind
+    TyVar _ _ | y : _ <- [ y | (y, k') <- Map.toList ctx
+                             , k == k' ] -> TyVar () y
+              | otherwise -> minimalType k
+    -- Try to fix application by fixing the function
+    TyApp _ a b       -> TyApp () (fixKind ctx a $ KindArrow () (inferKind_ ctx b) k) b
+    TyLam _ x kx b    ->
+      case k of
+        -- Fix lambdas to * by getting substituting a minimal type for the argument
+        -- and fixing the body.
+        Type{}        -> fixKind ctx (substClosedType x (minimalType kx) b) k
```

Comment:
@michaelpj 

> does this obey the invariant? if so why?

`fixKind` always generates a smaller type, because that type is going to have a smaller kind compared to the one of the original type. Here it is particularly clear how the resulting kind is smaller.